File: frontend/src/types/billing.ts
/**
 * The billing domain.
 *
 * Four records, related by id and never by copy: an **order** is what was
 * bought, an **invoice** is the demand for payment, a **payment** is money
 * moving, and a **refund** (with an optional **credit note**) is money moving
 * back. A real schema would be `orders`, `invoices`, `invoice_items`,
 * `payments`, `refunds`, `credit_notes`, and these shapes are written to map
 * onto it row for row.
 *
 * Every amount is `Money` — an integer in the currency's minor unit. See
 * `lib/money.ts` for why.
 */

/** An integer number of minor units (paise for INR). Never a float. */
export type Money = number;

export interface CurrencyConfig {
  code: string;
  symbol: string;
  locale: string;
  /** Minor units per major unit, as a power of ten. 2 for INR. */
  decimals: number;
}

/* ----------------------------------------------------------------------- tax */

export type TaxType = "GST" | "NONE";

/**
 * How tax is worked out.
 *
 * `originState` is where the business is registered. Indian GST splits into
 * CGST + SGST when the place of supply matches it and becomes a single IGST
 * when it does not, which is the only rule this models.
 *
 * **This is a configurable representation of GST, not a compliance
 * implementation.** Real tax treatment depends on HSN codes, exemptions,
 * composition schemes, reverse charge, place-of-supply rules and thresholds
 * that belong in a backend with an accountant behind it.
 */
export interface TaxConfig {
  enabled: boolean;
  taxType: TaxType;
  /** Whether catalogue prices already contain the tax. */
  pricesIncludeTax: boolean;
  /** The seller's state, for the intra/inter-state decision. */
  originState: string;
  /** Registration number shown on invoices. Display only. */
  gstin: string;
  rates: {
    cgst: number;
    sgst: number;
    igst: number;
  };
  /** Per-category overrides, by category slug. Falls back to `rates`. */
  categoryRates?: Record<string, { cgst: number; sgst: number; igst: number }>;
}

export type TaxMode = "intra-state" | "inter-state" | "none";

export interface TaxBreakdown {
  mode: TaxMode;
  /** The value tax was charged on, exclusive of tax. */
  taxableAmount: Money;
  cgst: Money;
  sgst: Money;
  igst: Money;
  totalTax: Money;
  /** The combined rate applied, for the "Tax (18%)" label. */
  ratePercent: number;
}

/* ------------------------------------------------------------------ breakdown */

/**
 * One money picture, used everywhere.
 *
 * The bag, every checkout step, the order record, the invoice, the admin order
 * page and the reports all render this same shape, produced by one function in
 * `billingService`. Components that each did their own arithmetic is how a
 * checkout ends up quoting two different totals.
 *
 * Components are preserved rather than collapsed into a total: an invoice has
 * to be able to show what was charged and why, and a refund has to be able to
 * work backwards through it.
 */
export interface BillingBreakdown {
  currency: string;
  itemCount: number;
  /** Line values at selling price, before order-level discounts. */
  subtotal: Money;
  /** Savings already reflected in selling prices, against list price. */
  productDiscount: Money;
  couponDiscount: Money;
  couponCode: string | null;
  shipping: Money;
  /** Anything the catalogue does not price — handling, gift wrap, COD fee. */
  otherCharges: Money;
  taxableAmount: Money;
  tax: TaxBreakdown;
  grandTotal: Money;
  /** True when `subtotal` already contains `tax.totalTax`. */
  pricesIncludeTax: boolean;
}

/* --------------------------------------------------------------- addresses */

/**
 * A billing address.
 *
 * Wider than the shipping `Address`: an invoice needs a country and a
 * contactable email, and the place of supply is read from `state`.
 */
export interface BillingAddress {
  fullName: string;
  phone: string;
  email: string;
  line1: string;
  line2: string;
  city: string;
  state: string;
  postalCode: string;
  country: string;
}

/* ----------------------------------------------------------------- invoices */

export type InvoiceStatus = "draft" | "issued" | "paid" | "overdue" | "cancelled";

export interface InvoiceLine {
  productId: string;
  name: string;
  sku: string;
  /** Harmonised System code. Placeholder data — see `TaxConfig`. */
  hsn: string;
  size: string | null;
  color: string | null;
  quantity: number;
  /** Selling price per unit, as charged. */
  unitPrice: Money;
  /** `unitPrice × quantity`. */
  lineSubtotal: Money;
  /** This line's share of order-level discounts. */
  discount: Money;
  taxableAmount: Money;
  taxRatePercent: number;
  cgst: Money;
  sgst: Money;
  igst: Money;
  tax: Money;
  /** What this line contributes to the grand total. */
  lineTotal: Money;
}

export interface Invoice {
  id: string;
  /** Human-facing, sequential, e.g. "DCZ-INV-2026-000001". */
  invoiceNumber: string;
  orderId: string;
  orderNumber: string;
  customerId: string;
  customerName: string;
  customerEmail: string;
  status: InvoiceStatus;
  issuedAt: string;
  dueAt: string;
  billingAddress: BillingAddress;
  shippingAddress: BillingAddress;
  /** The state tax was charged against. */
  placeOfSupply: string;
  lines: InvoiceLine[];
  breakdown: BillingBreakdown;
  paymentId: string | null;
  paymentMethod: PaymentMethodKey;
  paymentStatus: BillingPaymentStatus;
  /** Settled so far. Below `grandTotal` on a part-paid invoice. */
  amountPaid: Money;
  amountRefunded: Money;
  notes: string;
  terms: string;
}

/* ----------------------------------------------------------------- payments */

export type PaymentMethodKey =
  | "upi"
  | "card"
  | "debit-card"
  | "netbanking"
  | "cod"
  | "wallet";

export type BillingPaymentStatus =
  | "pending"
  | "authorized"
  | "paid"
  | "failed"
  | "refunded"
  | "partially-refunded";

export interface PaymentEvent {
  status: "initiated" | "processing" | "authorized" | "succeeded" | "failed" | "refunded";
  at: string;
  note: string;
}

export interface Payment {
  id: string;
  /** The reference a customer quotes to support, e.g. "TXN2026090412345". */
  transactionId: string;
  orderId: string;
  orderNumber: string;
  invoiceId: string;
  invoiceNumber: string;
  customerId: string;
  customerName: string;
  customerEmail: string;
  amount: Money;
  refundedAmount: Money;
  method: PaymentMethodKey;
  status: BillingPaymentStatus;
  /** Which integration handled it. "mock" until a real one is wired in. */
  provider: string;
  createdAt: string;
  capturedAt: string | null;
  /** Append-only. Drives the transaction timeline. */
  timeline: PaymentEvent[];
  /**
   * Safe, non-identifying remnants only — a card's last four digits, a masked
   * UPI handle. Never a full number, never a CVV, never a credential.
   */
  instrumentHint: string;
}

/* ------------------------------------------------------------------ refunds */

export type RefundStatus = "requested" | "processing" | "completed" | "rejected";

export interface RefundLine {
  productId: string;
  name: string;
  quantity: number;
  amount: Money;
}

export interface Refund {
  id: string;
  refundNumber: string;
  orderId: string;
  orderNumber: string;
  invoiceId: string;
  invoiceNumber: string;
  paymentId: string;
  customerId: string;
  customerName: string;
  amount: Money;
  reason: string;
  status: RefundStatus;
  requestedAt: string;
  processedAt: string | null;
  /** Empty for a whole-order refund; populated for a per-item one. */
  lines: RefundLine[];
  creditNoteId: string | null;
  /** Who initiated it — an admin id, or "customer". */
  initiatedBy: string;
}

/* ------------------------------------------------------------- credit notes */

export type CreditNoteStatus = "draft" | "issued" | "cancelled";

export interface CreditNote {
  id: string;
  creditNoteNumber: string;
  invoiceId: string;
  invoiceNumber: string;
  orderId: string;
  orderNumber: string;
  customerId: string;
  customerName: string;
  refundId: string | null;
  reason: string;
  /** Value before tax. */
  amount: Money;
  tax: Money;
  total: Money;
  issuedAt: string;
  status: CreditNoteStatus;
}

/* -------------------------------------------------------------- configuration */

export interface BillingConfig {
  business: {
    legalName: string;
    storeName: string;
    email: string;
    phone: string;
    website: string;
    addressLine1: string;
    addressLine2: string;
    city: string;
    state: string;
    postalCode: string;
    country: string;
  };
  currency: CurrencyConfig;
  invoice: {
    prefix: string;
    /** The first number a fresh sequence issues. */
    startNumber: number;
    /** Zero-padded width of the numeric part. */
    padding: number;
    /** Days from issue to due. */
    dueDays: number;
    footer: string;
    paymentTerms: string;
    notes: string;
  };
  creditNote: {
    prefix: string;
    startNumber: number;
    padding: number;
  };
  refund: {
    /** Days after delivery a refund can still be raised. Advisory here. */
    windowDays: number;
    /** Whether shipping is given back on a whole-order refund. */
    refundShipping: boolean;
    reasons: string[];
  };
  payment: {
    /** Methods offered at checkout, in order. */
    enabledMethods: PaymentMethodKey[];
    /** Extra charge for cash on delivery, in minor units. */
    codFee: Money;
  };
}

/* ----------------------------------------------------------------- reporting */

export interface BillingStats {
  revenue: Money;
  paid: Money;
  pending: Money;
  refunded: Money;
  taxCollected: Money;
  shippingRevenue: Money;
  discountsGiven: Money;
  invoiceCount: number;
  paymentCount: number;
  refundCount: number;
  creditNoteCount: number;
  /** Grand total minus refunds — what the business actually kept. */
  netSales: Money;
}

/** One row of the tax report: how much was collected at each rate. */
export interface TaxReportRow {
  ratePercent: number;
  taxableAmount: Money;
  cgst: Money;
  sgst: Money;
  igst: Money;
  totalTax: Money;
  invoiceCount: number;
}

/** What `placeOrder` needs to bill for, beyond the cart itself. */
export interface BillingContext {
  billingAddress: BillingAddress;
  shippingAddress: BillingAddress;
  paymentMethod: PaymentMethodKey;
  shipping: Money;
  otherCharges: Money;
}

File: backend/app/schemas/billing.py
"""Billing shapes.

Amounts are integers in the currency's minor unit throughout — the same
convention the database and the frontend use. A response that quietly switched
to rupees would be a units bug waiting in every consumer.
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import Field

from app.schemas.base import CamelModel


class TaxBreakdown(CamelModel):
    mode: str
    taxable_amount: int
    cgst: int
    sgst: int
    igst: int
    total_tax: int
    rate_percent: float


class BillingBreakdown(CamelModel):
    currency: str
    item_count: int
    subtotal: int
    product_discount: int
    coupon_discount: int
    coupon_code: Optional[str] = None
    shipping: int
    other_charges: int
    taxable_amount: int
    tax: TaxBreakdown
    grand_total: int
    prices_include_tax: bool


class InvoiceLineOut(CamelModel):
    product_id: str
    name: str
    sku: str
    hsn: str
    size: Optional[str] = None
    color: Optional[str] = None
    quantity: int
    unit_price: int
    line_subtotal: int
    discount: int
    taxable_amount: int
    tax_rate_percent: float
    cgst: int
    sgst: int
    igst: int
    tax: int
    line_total: int


class InvoiceOut(CamelModel):
    id: str
    invoice_number: str
    order_id: str
    order_number: str
    customer_id: str
    customer_name: str
    customer_email: str
    status: str
    issued_at: datetime
    due_at: datetime
    billing_address: dict
    shipping_address: dict
    place_of_supply: str
    lines: List[InvoiceLineOut]
    breakdown: BillingBreakdown
    payment_id: Optional[str] = None
    payment_method: str
    payment_status: str
    amount_paid: int
    amount_refunded: int
    notes: str
    terms: str

    @classmethod
    def from_model(cls, invoice) -> "InvoiceOut":
        return cls(
            id=invoice.id,
            invoice_number=invoice.invoice_number,
            order_id=invoice.order_id,
            order_number=invoice.order_number,
            customer_id=invoice.customer_id,
            customer_name=invoice.customer_name,
            customer_email=invoice.customer_email,
            status=invoice.status,
            issued_at=invoice.issued_at,
            due_at=invoice.due_at,
            billing_address=invoice.billing_address or {},
            shipping_address=invoice.shipping_address or {},
            place_of_supply=invoice.place_of_supply,
            lines=[InvoiceLineOut.model_validate(line) for line in invoice.items],
            breakdown=BillingBreakdown(
                currency=invoice.currency,
                item_count=invoice.item_count,
                subtotal=invoice.subtotal,
                product_discount=invoice.product_discount,
                coupon_discount=invoice.coupon_discount,
                coupon_code=invoice.coupon_code,
                shipping=invoice.shipping,
                other_charges=invoice.other_charges,
                taxable_amount=invoice.taxable_amount,
                tax=TaxBreakdown(
                    mode=invoice.tax_mode,
                    taxable_amount=invoice.taxable_amount,
                    cgst=invoice.cgst,
                    sgst=invoice.sgst,
                    igst=invoice.igst,
                    total_tax=invoice.total_tax,
                    rate_percent=float(invoice.tax_rate_percent),
                ),
                grand_total=invoice.grand_total,
                prices_include_tax=invoice.prices_include_tax,
            ),
            payment_id=invoice.payment_id,
            payment_method=invoice.payment_method,
            payment_status=invoice.payment_status,
            amount_paid=invoice.amount_paid,
            amount_refunded=invoice.amount_refunded,
            notes=invoice.notes,
            terms=invoice.terms,
        )


class PaymentEventOut(CamelModel):
    status: str
    note: str
    occurred_at: datetime = Field(serialization_alias="at")


class PaymentOut(CamelModel):
    id: str
    transaction_id: str
    order_id: str
    order_number: str
    invoice_id: str
    invoice_number: str
    customer_id: str
    customer_name: str
    customer_email: str
    amount: int
    refunded_amount: int
    method: str
    status: str
    provider: str
    created_at: datetime
    captured_at: Optional[datetime] = None
    timeline: List[PaymentEventOut] = []
    # A masked remnant only — never an instrument. See the Payment model.
    instrument_hint: str

    @classmethod
    def from_model(cls, payment) -> "PaymentOut":
        return cls(
            id=payment.id,
            transaction_id=payment.transaction_id,
            order_id=payment.order_id,
            order_number=payment.order_number,
            invoice_id=payment.invoice_id,
            invoice_number=payment.invoice_number,
            customer_id=payment.customer_id,
            customer_name=payment.customer_name,
            customer_email=payment.customer_email,
            amount=payment.amount,
            refunded_amount=payment.refunded_amount,
            method=payment.method,
            status=payment.status,
            provider=payment.provider,
            created_at=payment.created_at_utc,
            captured_at=payment.captured_at,
            timeline=[PaymentEventOut.model_validate(e) for e in payment.events],
            instrument_hint=payment.instrument_hint,
        )


class RefundLineOut(CamelModel):
    product_id: str
    name: str
    quantity: int
    amount: int


class RefundOut(CamelModel):
    id: str
    refund_number: str
    order_id: str
    order_number: str
    invoice_id: str
    invoice_number: str
    payment_id: str
    customer_id: str
    customer_name: str
    amount: int
    reason: str
    status: str
    requested_at: datetime
    processed_at: Optional[datetime] = None
    lines: List[RefundLineOut] = []
    credit_note_id: Optional[str] = None
    initiated_by: str

    @classmethod
    def from_model(cls, refund) -> "RefundOut":
        return cls(
            **{
                key: getattr(refund, key)
                for key in (
                    "id", "refund_number", "order_id", "order_number", "invoice_id",
                    "invoice_number", "payment_id", "customer_id", "customer_name",
                    "amount", "reason", "status", "requested_at", "processed_at",
                    "credit_note_id", "initiated_by",
                )
            },
            lines=[RefundLineOut.model_validate(line) for line in refund.items],
        )


class CreditNoteOut(CamelModel):
    id: str
    credit_note_number: str
    invoice_id: str
    invoice_number: str
    order_id: str
    order_number: str
    customer_id: str
    customer_name: str
    refund_id: Optional[str] = None
    reason: str
    amount: int
    tax: int
    total: int
    issued_at: datetime
    status: str


# ------------------------------------------------------------- requests


class RefundCreate(CamelModel):
    invoice_id: str
    amount: int = Field(gt=0)
    reason: str = Field(min_length=2)
    lines: Optional[List[dict]] = None
    status: str = "completed"


class RefundStatusUpdate(CamelModel):
    status: str


class CreditNoteCreate(CamelModel):
    invoice_id: str
    total: int = Field(gt=0)
    reason: str = Field(min_length=2)
    refund_id: Optional[str] = None
    status: str = "issued"


class CreditNoteStatusUpdate(CamelModel):
    status: str


class MarkPaidRequest(CamelModel):
    # Omitted settles the whole outstanding balance.
    amount: Optional[int] = None


class BillingStats(CamelModel):
    revenue: int
    paid: int
    pending: int
    refunded: int
    tax_collected: int
    shipping_revenue: int
    discounts_given: int
    invoice_count: int
    payment_count: int
    refund_count: int
    credit_note_count: int
    net_sales: int


class TaxReportRow(CamelModel):
    rate_percent: float
    taxable_amount: int
    cgst: int
    sgst: int
    igst: int
    total_tax: int
    invoice_count: int
